File: 프로그래머스/2/12924. 숫자의 표현/숫자의 표현.py
def solution(n):
    answer = 0
    dp = [0] * (n + 1)
    dp[1] = 1
    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + i
    left = 0
    right = 1

    while left < right and right<=n:
        now = dp[right] - dp[left]
        if now == n:
            answer += 1
            right +=1
            left += 1

        elif now > n:
            left += 1

        else:
            right += 1

    return answer

File: 프로그래머스/2/12924. 숫자의 표현/숫자의 표현.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(int n) {
    int answer = 0;
    vector<int>dp(n+1,0);
    for(int i=1;i<=n;i++){
        dp[i] = dp[i-1] + i;
    }
    
    int left = 0;
    int right = 1;
    while(left<=right && right<=n){
        if(dp[right]-dp[left] <n){
            right++;
        }
        else if(dp[right]-dp[left]>n){
            left++;
        }
        else{
            left++;
            right++;
            answer++;
        }
    }

    return answer;
}